<a href="https://colab.research.google.com/github/toupasana/Unsupervised-ML/blob/main/Copy_of_ZOMATO_RESTAURANT_CLUSTERING_AND_SENTIMENT_ANALYSIS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Problem Statement**

Zomato is an Indian restaurant aggregator and food delivery start-up founded by Deepinder Goyal and Pankaj Chaddah in 2008. Zomato provides information, menus and user-reviews of restaurants, and also has food delivery options from partner restaurants in select cities.

India is quite famous for its diverse multi cuisine available in a large number of restaurants and hotel resorts, which is reminiscent of unity in diversity. Restaurant business in India is always evolving. More Indians are warming up to the idea of eating restaurant food whether by dining outside or getting food delivered. The growing number of restaurants in every state of India has been a motivation to inspect the data to get some insights, interesting facts and figures about the Indian food industry in each city. So, this project focuses on analysing the Zomato restaurant data for each city in India.

The Project focuses on Customers and Company, you have  to analyze the sentiments of the reviews given by the customer in the data and made some useful conclusion in the form of Visualizations. Also, cluster the zomato restaurants into different segments. The data is vizualized as it becomes easy to analyse data at instant. The Analysis also solve some of the business cases that can directly help the customers finding the Best restaurant in their locality and for the company to grow up and work on the fields they are currently lagging in.

This could help in clustering the restaurants into segments. Also the data has valuable information around cuisine and costing which can be used in cost vs. benefit analysis

Data could be used for sentiment analysis. Also the metadata of reviewers can be used for identifying the critics in the industry. 

# **Attribute Information**

## **Zomato Restaurant names and Metadata**
Use this dataset for clustering part

1. Name : Name of Restaurants

2. Links : URL Links of Restaurants

3. Cost : Per person estimated Cost of dining

4. Collection : Tagging of Restaurants w.r.t. Zomato categories

5. Cuisines : Cuisines served by Restaurants

6. Timings : Restaurant Timings

## **Zomato Restaurant reviews**
Merge this dataset with Names and Matadata and then use for sentiment analysis part

1. Restaurant : Name of the Restaurant

2. Reviewer : Name of the Reviewer

3. Review : Review Text

4. Rating : Rating Provided by Reviewer

5. MetaData : Reviewer Metadata - No. of Reviews and followers

6. Time: Date and Time of Review

7. Pictures : No. of pictures posted with review

<b>Exploratory Data Analysis on Zomato Dataset</b><br>
 – Loading the Dataset <br>
 – Data Columns<br>
 – Duplicated function<br>
 – Information of data<br>
 – Reading column names<br>
 – Changing the column names<br>
 – Transformation in Data Type<br>
 – Changing cost to float<br>
 – Reading rate of dataset<br>
 – Removing from rate<br>
 – Adjust the column<br>
 – Encode the input variation<br>
 – Co-relation between different variations<br>
 – Column /name<br>
 – Data Visualization<br>
 – Table Booking Rate vs Rate<br>
 – Location<br>
 – Location vs Rating<br>
 – Restaurant type<br>
 – Gaussian rest type and rating<br>
 – Type of Service <br>
 – Type and Rating<br>
 – Cost of restaurant<br>
 – No.of restaurant in the location<br>
 – Type of Restaurant<br>
 – Most famous restaurant chains <br>


In [ ]:
#importing all the libraries which are requied to do EDA and visualization
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

plt.style.use('dark_background')

These are the libraries which I am using for EDA and Visualization.

* Pandas : a software library written for the Python programming language for data manipulation and analysis

* Numpy : a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays.

* Matplotlib : a plotting library for the Python programming language and its numerical mathematics extension NumPy.

* Seaborn : a Python data visualization library based on matplotlib. It provides a high-level interface for drawing attractive and informative statistical graphics.

### Loading the Dataset

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [7]:
restaurant = pd.read_csv("/content/drive/MyDrive/Capstone Project/Unsupervised Machine Learning /Zomato Restaurant names and Metadata.csv")
restaurant.head()
#review =pd.read_csv("/content/drive/MyDrive/Capstone Project/Unsupervised Machine Learning /Zomato Restaurant reviews.csv")

Name  \
0                 Beyond Flavours   
1                        Paradise   
2                        Flechazo   
3  Shah Ghouse Hotel & Restaurant   
4      Over The Moon Brew Company   

                                               Links   Cost  \
0  https://www.zomato.com/hyderabad/beyond-flavou...    800   
1  https://www.zomato.com/hyderabad/paradise-gach...    800   
2  https://www.zomato.com/hyderabad/flechazo-gach...  1,300   
3  https://www.zomato.com/hyderabad/shah-ghouse-h...    800   
4  https://www.zomato.com/hyderabad/over-the-moon...  1,200   

                                         Collections  \
0  Food Hygiene Rated Restaurants in Hyderabad, C...   
1                                Hyderabad's Hottest   
2                 Great Buffets, Hyderabad's Hottest   
3                             Late Night Restaurants   
4  Best Bars & Pubs, Food Hygiene Rated Restauran...   

                                            Cuisines  \
0  Chinese, Continental, Kebab, European, South I...   
1                     Biryani, North Indian, Chinese   
2       Asian, Mediterranean, North Indian, Desserts   
3  Biryani, North Indian, Chinese, Seafood, Bever...   
4  Asian, Continental, North Indian, Chinese, Med...   

                                             Timings  
0      12noon to 3:30pm, 6:30pm to 11:30pm (Mon-Sun)  
1                                     11 AM to 11 PM  
2              11:30 AM to 4:30 PM, 6:30 PM to 11 PM  
3                                    12 Noon to 2 AM  
4  12noon to 11pm (Mon, Tue, Wed, Thu, Sun), 12no...

In [11]:
restaurant.shape

(105, 6)

We can see that there are 105 rows of Data and 6 Columns

In [10]:
restaurant.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105 entries, 0 to 104
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Name         105 non-null    object
 1   Links        105 non-null    object
 2   Cost         105 non-null    object
 3   Collections  51 non-null     object
 4   Cuisines     105 non-null    object
 5   Timings      104 non-null    object
dtypes: object(6)
memory usage: 5.0+ KB


When we looked into the information of the data we can see some values are missing in different columns Collection,timing(for one)



In [12]:
#Checking the nullm values in dataset
restaurant.isna().sum()

Name            0
Links           0
Cost            0
Collections    54
Cuisines        0
Timings         1
dtype: int64

In [13]:
# Lets find the total number of null values

restaurant.isnull().sum().sum()
print('The total number of null values in the whole dataset is {}.'.format(restaurant.isnull().sum().sum()))

The total number of null values in the whole dataset is 55.


In [15]:
#print('Convert the 'Cost' column, deleting the comma and changing the data type into 'int64'')
restaurant['Cost'] = restaurant['Cost'].str.replace(",","").astype('int64')

In [18]:
#Check the all data info
restaurant.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105 entries, 0 to 104
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Name         105 non-null    object
 1   Links        105 non-null    object
 2   Cost         105 non-null    int64 
 3   Collections  51 non-null     object
 4   Cuisines     105 non-null    object
 5   Timings      104 non-null    object
dtypes: int64(1), object(5)
memory usage: 5.0+ KB


In [19]:
# Examine Statistics
restaurant.describe()

Cost
count   105.000000
mean    861.428571
std     510.149730
min     150.000000
25%     500.000000
50%     700.000000
75%    1200.000000
max    2800.000000

In [20]:
#find the most popular 5 collections
import itertools
import collections
from collections import Counter
import operator
restaurant_collections=restaurant['Collections'].dropna()
list_of_collections=list(itertools.chain.from_iterable(restaurant_collections.str.split(',').to_list()))
list_of_collections=list(map(str.strip,list_of_collections))
top_5_collections=[a[0] for a in Counter(list_of_collections).most_common(5)]

In [21]:
print(top_5_collections)

['Great Buffets', 'Food Hygiene Rated Restaurants in Hyderabad', 'Live Sports Screenings', "Hyderabad's Hottest", 'Corporate Favorites']


In [22]:
#find the most popular and least popular 5 cuisines
import itertools
import collections
from collections import Counter
import operator
restaurant_cuisines=restaurant['Cuisines'].dropna()
list_of_cuisines=list(itertools.chain.from_iterable(restaurant_cuisines.str.split(',').to_list()))
list_of_cuisines=list(map(str.strip,list_of_cuisines))
#print(list_of_cuisines)
popular_cuisines=[a[0] for a in Counter(list_of_cuisines).most_common()][:5]
print(popular_cuisines)
count_dict_cuisines=dict(Counter(list_of_cuisines))
order_by_pop_cusines=collections.OrderedDict(sorted(count_dict_cuisines.items(),key=lambda x:x[1]))

least_popular_cuisines=[a for a in order_by_pop_cusines][:10]
print(least_popular_cuisines)


['North Indian', 'Chinese', 'Continental', 'Biryani', 'Asian']
['Goan', 'Healthy Food', 'Lebanese', 'Mexican', 'Juices', 'Spanish', 'Wraps', 'Indonesian', 'Pizza', 'North Eastern']
